File: apzkr-pzpi-21-11-milennyi-ivan/Task3-WebClient/src/api/services/StorekeeperService.ts
import axios from "axios";
import { Storekeeper } from "../../models/Storekeeper";
import { API_URL } from "../../utils/config";
import { AuthService } from "../interfaces/IAuthService";
import { IStorekeeperService } from "../interfaces/IStorekeeperService";

export class StorekeeperService implements IStorekeeperService, AuthService<Storekeeper> {
    // Статичні URL-адреси для запитів до API
    private static STOREKEEPER_URLS = {
        GET_ALL: `${API_URL}/storekeeper`, // URL для отримання всіх комірників
        GET_BY_ID: (id: number) => `${API_URL}/storekeeper/${id}`, // URL для отримання комірника за ідентифікатором
        CREATE: `${API_URL}/storekeeper/create`, // URL для створення нового комірника
        DELETE: (id: number) => `${API_URL}/storekeeper/delete/${id}`, // URL для видалення комірника за ідентифікатором
        UPDATE: `${API_URL}/storekeeper/update`, // URL для оновлення комірника
        AUTHORIZE: `${API_URL}/storekeeper/authorize` // URL для авторизації комірника
    }

    // Метод для створення нового комірника
    async create(item: Storekeeper): Promise<Storekeeper> {
        const response = await axios.post<Storekeeper>(StorekeeperService.STOREKEEPER_URLS.CREATE, item);
        return response.data;
    }

    // Метод для видалення комірника за ідентифікатором
    async delete(itemId: number): Promise<void> {
        await axios.delete(StorekeeperService.STOREKEEPER_URLS.DELETE(itemId));
    }

    // Метод для оновлення існуючого комірника
    async update(item: Storekeeper): Promise<Storekeeper> {
        const response = await axios.put<Storekeeper>(StorekeeperService.STOREKEEPER_URLS.UPDATE, item);
        return response.data;
    }

    // Метод для отримання всіх комірників
    async getAll(): Promise<Storekeeper[]> {
        const response = await axios.get<Storekeeper[]>(StorekeeperService.STOREKEEPER_URLS.GET_ALL);
        return response.data;
    }

    // Метод для отримання комірника за ідентифікатором
    async getById(id: number): Promise<Storekeeper | null> {
        const response = await axios.get<Storekeeper | null>(StorekeeperService.STOREKEEPER_URLS.GET_BY_ID(id));
        return response.data;
    }

    // Метод для авторизації комірника
    async authorize(username: string, passwordHash: string): Promise<Storekeeper | null> {
        const response = await axios.post<Storekeeper | null>(StorekeeperService.STOREKEEPER_URLS.AUTHORIZE, { username: username, password_hash: passwordHash });
        return response.data;
    }
}

File: apzkr-pzpi-21-11-milennyi-ivan/Task1-Server/src/db/services/storekeeper_service.rs
use std::sync::Arc;
use async_trait::async_trait;
use sqlx::{MySql, Pool, query, query_as};
use crate::db::service_error::ServiceError;
use crate::db::traits::{AuthService, Service, StorekeeperManage};
use crate::models::Storekeeper;
use crate::view_models::StorekeeperVM;

// Сервіс для управління комірниками
pub(crate) struct StorekeeperService<T>{
    pool: Arc<T>,
}

#[async_trait]
impl Service<Pool<MySql>> for StorekeeperService<Pool<MySql>> {
    type Model = Storekeeper;
    type Error = ServiceError;
    type ViewModel = StorekeeperVM;

    fn new(pool: Arc<Pool<MySql>>) -> Self {
        StorekeeperService { pool }
    }

    // Функція для створення нового комірника
    async fn create(&self, mut item: Self::Model) -> Result<Self::Model, Self::Error> {
        query(
            r#"
            INSERT INTO Storekeepers (username, password, name, surname)
            VALUES (?, ?, ?, ?)
            "#
        )
            .bind(item.username())
            .bind(item.password())
            .bind(item.name())
            .bind(item.surname())
            .execute(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
            .map(|result|
                if result.rows_affected() == 1 {
                    item.set_id(result.last_insert_id());
                    Ok(item)
                }
                else{
                    Err(ServiceError::CustomError("Insertion went wrong. Zero rows affected".to_string()))
                }
            )
            .unwrap_or_else(|error| Err(error))
    }

    // Функція для видалення комірника за його ідентифікатором
    async fn delete(&self, item_id: u64) -> Result<(), Self::Error> {
        query(
            r#"
            DELETE FROM Storekeepers
            WHERE id = ?
            "#
        )
            .bind(item_id)
            .execute(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
            .map(|result|
                if result.rows_affected() == 0 {
                    Err(ServiceError::CustomError("Zero rows affected".to_string()))
                }
                else{
                    Ok(())
                }
            )
            .unwrap_or_else(|error| Err(error))
    }

    // Функція для оновлення інформації про комірника
    async fn update(&self, item: Self::Model) -> Result<Self::Model, Self::Error> {
        query(
            r#"
            UPDATE Storekeepers
            SET username = ?, password = ?, name = ?, surname = ?
            WHERE id = ?
            "#
        )
            .bind(item.username())
            .bind(item.password())
            .bind(item.name())
            .bind(item.surname())
            .bind(item.id().ok_or(ServiceError::CustomError("ID is required".to_string()))?)
            .execute(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
            .map(|result|
                if result.rows_affected() == 0 {
                    Err(ServiceError::CustomError("Zero rows affected".to_string()))
                }
                else{
                    Ok(item)
                }
            )
            .unwrap_or_else(|error|  Err(error))
    }

    // Функція для отримання всіх комірників
    async fn get_all(&self) -> Result<Vec<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT * FROM Storekeepers
            "#
        )
            .fetch_all(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
    }

    // Функція для отримання комірника за його ідентифікатором
    async fn get_by_id(&self, id: u64) -> Result<Option<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT * FROM Storekeepers
            WHERE id = ?
            "#
        )
            .bind(id)
            .fetch_optional(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
    }
}

#[async_trait]
impl AuthService<Pool<MySql>> for StorekeeperService<Pool<MySql>> {
    // Функція для перевірки унікальності імені користувача
    async fn check_username(&self, user: &Self::Model) -> Result<bool, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT * FROM Storekeepers
            WHERE username = ?
            "#
        )
            .bind(user.username())
            .fetch_optional(&*self.pool).await
            .map(|result| result.is_some()).map_err(|error| ServiceError::DatabaseError(error))
    }

    // Функція для авторизації користувача
    async fn authorize(&self, username: String, password_hash: String) -> Result<Option<Self::ViewModel>, Self::Error> {
        query_as::<_, Self::ViewModel>(
            r#"
            SELECT id, name, surname FROM Storekeepers
            WHERE username = ? AND password = ?
            "#
        )
            .bind(username)
            .bind(password_hash)
            .fetch_optional(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
    }
}

#[async_trait]
impl StorekeeperManage<Pool<MySql>> for StorekeeperService<Pool<MySql>>{

}
